Implement basic support for reader conditionals

vemv opened the pull request:
Hey there,

here's my stab at https://github.com/gfredericks/how-to-ns/issues/9 .

Probably interspersing code with `reader-conditional?` isn't very elegant, so I'm open to change it to sth else. Thought I'd be cautious to start small :)

Strictly speaking , reader conditionals can appear anywhere in the code, but this PR only deals with likely choices from the user.

There's no attempt whatsoever at formatting the contents of the reader conditional. Can eventually do.

WDYT?

Victor

vemv committed:
Implement basic support for reader conditionals

vemv committed:
Support npm-style requires

Fixes #11

vemv committed:
Implement `:place-string-requires-at-bottom?` option

gfredericks commented on how-to-ns/src/com/gfredericks/how_to_ns.clj:
```diff
@@ -99,24 +99,41 @@
          (group-by first)
-         (sort)
+         (sort-by (fn [[x]]
+                    (str x)))
```
`x` might be a reader conditional here, is that correct?

according to my repl, `str` on a reader conditional gives something like `clojure.lang.ReaderConditional@5044f1cd`, which seems like it would yield a nondeterministic sort here; `pr-str` would likely do better; I'd like something that involves the contents of the conditional, but I don't know the best way to do that; if you know what order the kv-pairs inside it would come in, you could sort based on the first value.

```

vemv replied:
> x might be a reader conditional here, is that correct?

I scienced it and yes, correct. Will be fixing

gfredericks reviewed (commented):
I mentioned a sorting issue or two, and I also started thinking about the align-clauses issue, and how weird it will look to have the reader conditionals sprinkled throughout the otherwise-aligned clauses.

Note that you can use the `:form` and `:splicing?` keywords to extract the data parts of a reader conditional, and the `reader-conditional` function to reconstruct one.

This means we could possibly define some kind of canonical ordering for mixing reader conditionals and regular clauses, e.g. by first sorting the pairs inside the reader conditional, sorting reader conditionals based on their sequences of pairs, and placing all reader conditionals before (or after) vanilla clauses.

I'm not sure it's worth trying to define alignment for reader conditionals, except maybe if there are multiple clauses inside of splicing conditionals.

vemv committed:
Rename: `:place-string-requires-at-bottom?` -> `:sort-string-requires-to-end?`

vemv committed:
`require-sort-criterion`: sort more cleanly

vemv committed:
Merge branch 'npm' into read-cond

# Conflicts:
#	how-to-ns/src/com/gfredericks/how_to_ns.clj

vemv committed:
Handle parenthesized imports with just one member

vemv committed:
Further handle reader conditionals within :import clauses

vemv committed:
Prefer pr-str to str as a sort criterion

vemv commented:
Added support for various corner cases (reader conditionals placed in presumably unusual places), covered by tests and with per-commit granularity.

Turns out, supporting those could be as easy as upgrading `str` to `pr-str`, or similar minor tweaks.

Also fixed the corner case of `(:import (Thing))`, i.e. the `(Thing)` list having just one member. It was buggy in master even w/o reader conditionals.

Note that this branch has `npm` merged in - it will become cleaner after we are done with https://github.com/gfredericks/how-to-ns/pull/12

gfredericks commented:
Seeing as

1. this problem is super complicated
2. I don't have immediate plans to make use of this feature
3. and I'm pretty busy
4. the prior two points mean I'm not very motivated to give it a lot of attention
5. but having the library not work at all when reader conditionals are around greatly reduces its utility

I'm going to merge this and add a disclaimer to the README saying that behavior with reader conditionals is subject to change.

Thanks for the work!

gfredericks commented:
Released as `0.2.5`, thanks again.
